### DISCUSSION TITLE
Autosuggestions

### DISCUSSION THREAD

**Discussion by leo-arch** [leo-arch/clifm#47] [Show and tell]
Hi @tuerda, @Chiheisen, @sundaran1122, @IcyFlamingArrow, @comradekingu, @jeeger . I've been working on a [Fish-like suggestions system](https://github.com/leo-arch/clifm/wiki/Specifics#auto-suggestions) for CliFM for the last two weeks, and I would love to have some users feedback. Today I can say that it is pretty stable and works as intended, but it's still a very young feature and, as such, needs more testing.

It is not enabled by default yet, so that you need to explicitly enable it via the `--suggestions` command line option, or setting `AutoSuggestions` to `true` in the configuration file. Btw, a few [new options](https://github.com/leo-arch/clifm/blob/master/misc/clifmrc) (mostly for suggestions) have been added to the configuration file.

Just in case you're interested, the code for this feature is in `src/suggestions.c`.

As always, your suggestions will be highly appreciated.
Thank you very much in advance.

**Comment by tuerda** [comment]
Sorry took so long to see this. Github writes emails to an address I check rarely.

Anyway, it seems to work. This is not a feature I personally need or want, but it looks like it does what it is supposed to.

**Reply by leo-arch** [reply]
Hi @tuerda . Don't worry. Still working on it. Just let me know if some issue comes out. I want this feature to be quite stable before releasing it in the next version (v1.2). I'm planning to make it the default.

Do you think it lacks something, that suggestions for something else could be added? I already added suggestions for ELN's, bookmarks, aliases, parameters for internal commands, and the jump function. Quite useful when it comes to ELN's and when jumping directories with the `j` command, imho. Otherwise, I feel like entering commands a bit blindly; sorta, is this the ELN I want, am I jumping to the right directory?

**Reply by tuerda** [reply]
I am not the right person to ask, unfortunately.
I never use autocompletion anywhere for any reason.
I like my manual tab completion, but I find automatic version to be a source of consant visual noise, and I turn it off everywhere I possibly can.

I tried this and it did what you said it did.
I didn't find any bugs or weird issues.

I used it for about 5-10 minutes just to test it and play around with it; then I turned it back off.
It is an _extremely_ popular feature almost everywhere, so it probably is a good idea to include it, but it just isn't something I would use myself.

I guess the ELN thing is a good point, but when using the ELNs, I have already looked up the number in the list that was provided, so I dunno . . . potentially pretty great, but not for me.

I am the weirdo here though. Almost everybody else thinks autocompletion is the best thing since sliced bread.

**Reply by leo-arch** [reply]
Yeap, I guess you are a bit of a weirdo. But not because of the suggestions stuff: CliFM is after all already for weirdos: almost everyone out there do not even consider an alternative to TUI file managers. In this case, I'm the king of weirdos; so, don't worry, you're not alone.

As to suggestions, it happened to me to delete the wrong file just because the ELN I read was not actually the one I wanted (add to this the fact that ELN's might be dangerous when using concatenated commands, e.g. `r1 && r2`). At least for me, suggestions is a way to prevent people (including myself) from committing this kind of stupid mistakes. Of course, TAB completion also helps with this.

Be it as it may, thank you very much for sharing your point of view. It is always helpful to know users opinions, no matter what they are.

**Comment by Chiheisen** [comment]
Hi @leo-arch 

Missed the first ping somehow, only just saw the newer notifications. Def. a feature I'm happy about having and that I'll keep turned on.
These are just initial thoughts, I'll have a more complete feedback after I've used it more. 
Works really well for what I've thrown at it. It is quite nice for ELNs for some of my use cases, especially when dealing with large folders (which contain mostly sequentially numbered data) where I about know where it should be, but want a confirmation first without having to tab, and then delete if it was the wrong thing. Faster feedback loop is always a good thing IME.

I've found one issue: History seems to not consistently work for me. It works on `cd`, `bat`, `mpv` (and many more) but not `cat` or `nvim`, `emacsclient -t`. Haven't been able to find a very consistent way to repro this so I'm not sure what the cause is. 
Other than that, so far no complaints, only praise.

**Reply by leo-arch** [reply]
Hi @Chiheisen. Don't worry about the missed notification and thanks for this feedback.

I'll add some random entries to my history using the commands you pointed out to try to get some unexpected behavior (I already observed some weird behaviors with history while testing on BSD, but wasn't sure). Provided some issue comes out, and I'm sure it will, it should be quite trivial to solve.

I'll let you know.

**Reply by leo-arch** [reply]
It should be working as expected now. Try it out and let me know.

**Reply by Chiheisen** [reply]
Fast as always! Yes, looks like it works as I'd expect now.

**Reply by leo-arch** [reply]
Great! Please let me know if you find some other issue, or have some suggestion on how to improve this feature.
Thanks @Chiheisen.
